Multitouch Support/Native/VoodooI2CNativeEngine.cpp

```diff
@@ -70,11 +72,60 @@ MultitouchReturn VoodooI2CNativeEngine::handleInterruptReport(VoodooI2CMultitouc
         }
     }
     
+    // Reset finger tracking
+    if (allInvalid) {
+        memset(freeFingerTypes, true, kMT2FingerTypeCount);
+        freeFingerTypes[kMT2FingerTypeUndefined] = false;
+    }
+    
+    // Fixes 2F taps as MT stack
+    if (event.contact_count > 2) {
+        setThumbFingerType(&message);
+    }
+    
+    for(int i = 0; i < event.contact_count; i++) {
+        VoodooInputTransducer* inputTransducer = &message.transducers[i];
+        
+        if (inputTransducer->fingerType == kMT2FingerTypeUndefined) {
```

Comment:
always true because of the assignment above?
```cpp
inputTransducer->fingerType = kMT2FingerTypeUndefined;
```

It will fail to preserve the previous set finger type.

Reply:
@Goshin as I am not familiar with the problem, would you mind taking an attempt at it?